Rocky.class.nut

```diff
@@ -199,50 +245,84 @@ class Rocky {
         if (contentType.find("multipart/form-data;") == 0) {
             local parts = [];
 
-            // _parse_body is wrapped in a try/catch.. so we just let this fail
-            // when the content-type isn't long enough (and on other issues).
-            local boundary = contentType.slice(30);
+            // Find the boundary in the contentType
+            local boundary;
+            local findString = regexp(@"boundary([ ]*)=");
+            local match = findString.search(contentType);
+
+            if (match) {
+                boundary = contentType.slice(match.end);
+                boundary = strip(boundary);
+            } else {
+                throw "No boundary found in content-type";
+            }
+
+            // Remove all carriage returns from string (to support either \r\n or \n for linebreaks)
+            local body = "";
+            local bodyLines = split(req.body, "\r");
+            foreach (i, line in bodyLines) {
+                body += line;
+            }
 
             local bindex = -1;
             do {
-                bindex = req.body.find("--" + boundary + "\r\n", bindex+1);
+                bindex = body.find("--" + boundary + "\n", bindex+1);
 
                 if (bindex != null) {
                     // Locate all the parts
-                    local hstart = bindex + boundary.len() + 4;
-                    local nstart = req.body.find("name=\"", hstart) + 6;
-                    local nfinish = req.body.find("\"", nstart);
-                    local fnstart = req.body.find("filename=\"", hstart) + 10;
-                    local fnfinish = req.body.find("\"", fnstart);
-                    local bstart = req.body.find("\r\n\r\n", hstart) + 4;
-                    local fstart = req.body.find("\r\n--" + boundary, bstart);
-
-                    // Pull out the parts as strings
-                    local headers = req.body.slice(hstart, bstart);
+                    local hstart = bindex + boundary.len() + 3;
+                    local hfinish = body.find("\n\n", hstart);
+                    local header = body.slice(hstart, hfinish);
+
+                    // Get the name
                     local name = null;
+                    local nstart = header.find(" name=\"");
```

Comment:
This code would look much more elegant if you use regexps. Is there a specific reason not to use them?

Reply:
The current manual parsing (vs a single regular expression per the header to extract all the groups):

a) blows up the code size
b) is sub-efficient
c) is error-prone

If we eventually end up leaving the "manual" regular expression work, it needs to be heavily unit-tested.

Reply:
in fact you can use just one regular expression to extract all these components. there is no need to do it all manually. and if there is a need to extract them manually then I would at least extract this into a separate function to avoid code duplication.

Reply:
The reason I didn't use regexp is because there is no way in squirrel's implementation to match multiple instances in one string. It will only return the first match. This is a problem because of there being multiple files in one body, meaning we need to match multiple instances of these fields. A potential way to use regexp if you would prefer would be to split the body on the boundary string and then use regexp separately on each split section. What would you like me to do?

Reply:
splitting the body means creating an array that contains all the response. while I believe replacing the internal manual parsing for the content inside the `do` loop with regexp would already make it more readable, simpler and reliable. does it make sense?